File: examples/borrow_checker1.cpp
#include <iostream>


class Resource {
private:
    int* data;

public:
    Resource(int data) {
        this->data = new int;
        *this->data = data;
    }
    ~Resource() {
        delete this->data;
    }

    const int* getData() const {
        return this->data;
    }
};

int main() {
    const Resource foo {0};
    const int* data = foo.getData();
    std::cout << *data << "\n";

    foo.~Resource(); // very obscure

    const int copy = *data; // UB !!
    std::cout << copy  << "\n";

    // Note: foo.~Resource(); is also
    // automatically called here, which is
    // actually a double-free, UB yet again!
}

File: examples/borrow_checker1.rs
use std::mem::{drop};

struct Resource {
    data: Box<i32>,
}

impl Resource {
    fn new(data: i32) -> Self {
        Self { data: Box::new(data) }
    }

    fn data(&self) -> &i32 {
        &self.data
    }
}

fn main() {
    let foo = Resource::new(0);
    let data = foo.data();

    println!("{data}");

    drop(foo);

    let copy = *data; // Compile Error!!
    println!("{copy}");

    // No double-free in Rust!
    // This function no longer owns foo
    // so it is no longer responsible for
    // free-ing it!
}